ddtrace/tracer.py

```diff
@@ -452,3 +458,70 @@ def set_tags(self, tags):
         :param dict tags: dict of tags to set at tracer level
         """
         self.tags.update(tags)
+
+    def on(self, hook_name):
+        """
+        Decorator used to configure a tracer span hook
+
+        Usage::
+
+            # Register a hook for any emitted `my.span` Spans
+            @tracer.on('my.span')
+            def on_falcon_request(span, data):
+                span.set_tag('key', data['key'])
+
+
+            # Create and emit a `my.span` Span with the provided data
+            with tracer.trace('my.span') as span:
+                tracer.emit(span, dict(key='value'))
+
+        :param hook_name: The name of the hook to configure a hook for
+        :type hook_name: str
+        :rtype: function
+        :returns: A function decorator that registers a function for the provided hook name
+        """
+        def wrapper(func):
+            self.hooks[hook_name].add(func)
+            return func
+        return wrapper
+
+    def emit(self, span, *args, **kwargs):
+        """
+        Call all registered span hooks for the provided :class:`ddtrace.span.Span`
+
+        Usage::
+
+            # Register a hook for any emitted `my.span` Spans
+            @tracer.on('my.span')
+            def on_falcon_request(span, data):
+                span.set_tag('key', data['key'])
+
+
+            # Create and emit a `my.span` Span with the provided data
+            with tracer.trace('my.span') as span:
+                tracer.emit(span, dict(key='value'))
+
+        :param span: The span to emit a hook for
+        :type span: :class:`ddtrace.span.Span`
+        :param *args: positional arguments to pass to hook functions
+        :param **kwargs: keyword arguments to pass to hook functions
+        :rtype: None
+        :raises: TypeError
+        """
+        if not isinstance(span, Span):
+            raise TypeError('Tracer.emit expected a `ddtrace.span.Span` instead got "{}"'.format(span))
+
+        # Return early if no hooks exist
+        # DEV: Do this check even though we use a `defaultdict` to save from
+        #      creating an entry for the name if one does not already exist
+        if span.name not in self._hooks:
+            return
+
+        # Iterate and call all hooks for this span name
+        for func in self._hooks[span.name]:
+            fname = func_name(func)
+            log.debug('calling "{}" hook function "{}"'.format(span.name, fname))
+            try:
+                func(span, *args, **kwargs)
+            except Exception as err:
+                log.error('error while calling "{}" hook function "{}": {}'.format(span.name, fname, err))
```

Comment:
We don't have a good way to throttle logging. If we use `log.error` (that is semantically correct) we risk to write 1k log entries per second if we hit a bug and this is in users' code critical path.

I think we should keep it as `log.debug` until we throttle our logging (i.e. not more than 5 log entries in 10 seconds).